### PR TITLE
[ARM] Stop gluing FP comparisons to FMSTAT

### PR DESCRIPTION
**Pull request by s-barannikov**
Following #116547, this changes the result of `ARMISD::CMPFP*` and the operand of `ARMISD::FMSTAT` from a special `Glue` type to a normal type.

This change allows comparisons to be CSEd and scheduled around as can be seen in the test changes.

Note that `ARMISD::FMSTAT` is still glued to its consumer nodes; this is going to be changed in a separate patch.

This patch also sets `CopyCost` of `cl_FPSCR_NZCV` register class to a negative value. The reason is the same as for CCR register class: it makes DAG scheduler and InstrEmitter try to avoid copies of `FPCSR_NZCV` register to / from virtual registers. Previously, this was not necessary, since no attempt was made to create copies in the first place.

There might be a case when a copy can't be avoided (although not found in existing tests). If a copy is necessary, the virtual register will be created with `cl_FPSCR_NZCV` register class. If this register class is inappropriate, `TRI::getCrossCopyRegClass` should be modified to return the correct class.

### CODE REVIEW

**Comment by statham-arm** [comment]
It looks as if the `FMSTAT` is no longer glued to the `VCMP` before it, but _is_ still glued to whatever uses its output PSR flags? (Based on the fact that `duplicateCmp` is still generating an `MVT::Glue` as output, and indeed needs to exist at all.)

Not a complaint – I don't want to be one of those code reviewers who won't accept a small improvement until it's turned into a much bigger one :-) but it might be a thing to call out in the commit message to avoid confusion.

**Comment by s-barannikov** [comment]
> It looks as if the `FMSTAT` is no longer glued to the `VCMP` before it, but _is_ still glued to whatever uses its output PSR flags? (Based on the fact that `duplicateCmp` is still generating an `MVT::Glue` as output, and indeed needs to exist at all.)

That's true, and I was going to fix it in the next patch. Unfortunately the patch will not be small, because changing this particular glue requires changing operands of BRCOND/CSINC/CSEL and all the others simultaneously. Otherwise, I'd have to insert copies and that would break DAG combiner logic.

> Not a complaint – I don't want to be one of those code reviewers who won't accept a small improvement until it's turned into a much bigger one :-) but it might be a thing to call out in the commit message to avoid confusion.

Sure, there are a couple more things I would also like to point out.

**Comment by s-barannikov** [comment]
The test failure on the Windows bot appears to be caused by #113697.

File: llvm/test/CodeGen/ARM/fptosi-sat-scalar.ll [hunk 1]
```diff
@@ -1108,51 +1064,24 @@ define i128 @test_signed_i128_f32(float %f) nounwind {
...
+; VFP-NEXT:    movgt.w r2, #-1
+; VFP-NEXT:    movgt.w r1, #-1
+; VFP-NEXT:    movgt.w r0, #-1
 ; VFP-NEXT:    vcmp.f32 s2, s2
 ; VFP-NEXT:    vmrs APSR_nzcv, fpscr
-; VFP-NEXT:    it vs
+; VFP-NEXT:    itttt vs
+; VFP-NEXT:    movvs r0, #0
+; VFP-NEXT:    movvs r1, #0
+; VFP-NEXT:    movvs r2, #0
```
Comment (statham-arm) [inline llvm/test/CodeGen/ARM/fptosi-sat-scalar.ll:1084]: Wow, this kind of thing is a _particularly_ impressive improvement!

**Review by statham-arm — APPROVED** [review]
> Sure, there are a couple more things I would also like to point out.

Ah, I see you've also explained the negative `CopyCost`. Great, I was wondering about that too :-)

**Comment by mstorsjo** [comment]
This change caused miscompilations (crashes due to illegal instruction) at runtime. It can also be observed as failed asserts if compiling to assembly output, like this:
```c
double a, b;
double c() {
  double d = a = b;
  if (d)
    a = -b;
  a = c();
  if (d)
    a = -a;
  return a;
}
```
```console
$ clang -target armv7-linux-gnueabihf -S repro.c -O2
clang: ../lib/Target/ARM/MCTargetDesc/ARMInstPrinter.cpp:1296: void llvm::ARMInstPrinter::printT2AddrModeImm8s4Operand(const llvm::MCInst*, unsigned int, const 
llvm::MCSubtargetInfo&, llvm::raw_ostream&) [with bool AlwaysPrintImm0 = false]: Assertion `((OffImm & 0x3) == 0) && "Not a valid immediate!"' failed.
```
(Note that compiling with `-c` doesn't trigger any errors, while compiling with `-S` does.)

The original, non-reduced case is also available at https://martin.st/temp/lfg-preproc.c, in case this reduction misses the original use case (the original case didn't involve a function calling itself I think).

Please have a look, or revert if fixing takes a while.

**Comment by s-barannikov** [comment]
Hi
Thanks for the report and analysis, I'll look right now. If the reason is not obvious, I'll revert the patch.

**Comment by s-barannikov** [comment]
The issue is that FPSCR gets spilled and restored, and the load and the store have incorrect addressing mode:

```
*** Bad machine code: Incorrect AddrMode Imm for instruction ***
- function:    c
- basic block: %bb.0 entry (0x5bd7d62ac2e0)
- instruction: VSTR_FPSCR_NZCVQC_off $fpscr_nzcv, $sp, 1, 14, $noreg :: (store (s32) into %stack.0)

*** Bad machine code: Incorrect AddrMode Imm for instruction ***
- function:    c
- basic block: %bb.0 entry (0x5bd7d62ac2e0)
- instruction: $fpscr_nzcv = VLDR_FPSCR_NZCVQC_off $sp, 1, 14, $noreg :: (load (s32) from %stack.0)
```

This can be fixed by handling the corresponding register class in `TRI::getCrossCopyRegClass`, but I'll need some time investigate it more thoroughly. Reverting for now.

**Comment by s-barannikov** [comment]
The machine verifier error above is wrong, the immediate *is* valid. `verifyInstruction` calls `isLegalAddressImm` with *encoded* immediate, while the function seems to expect decoded (scaled by 4) value.

The instruction have `t2addrmode_imm7s4` operands. Do I understand correctly that these instructions do not exist in ARM mode?
CC @ostannard

**Comment by mstorsjo** [comment]
> The machine verifier error above is wrong, the immediate _is_ valid. `verifyInstruction` calls `isLegalAddressImm` with an encoded immediate, while the function seems to expect a decoded (scaled by 4) value.
> 
> These instruction have `t2addrmode_imm7s4` operands. Do I understand correctly that these instructions do not exist in ARM mode? CC @ostannard

Hmmm - but when running the generated code, execution fails with Illegal instruction, both for ARM and Thumb mode. So even if the immediate may be valid, there's something wrong with the generated code here still...

**Comment by s-barannikov** [comment]
It looks like these instructions do not exist in ARM mode. At least when I fix the encoding issue, the resulting assembly file is not accepted by the assembler:
```
test.s:1:1: note: instruction requires: armv8.1m.main
vstr fpscr_nzcvqc, [sp, #4]
```

The encoding of this addressing mode is also broken, but I believe only in ARM mode.

NB
I'm not really familiar with ARM/Thumb modes, their cooperation, how they relate to processor models, which of them backward or forward compatible with which etc., so the above may be complete nonsense. I would love some help here.

**Comment by s-barannikov** [comment]
PR with candidate fix: https://github.com/llvm/llvm-project/pull/117248

It successfully compiles the test attached above, and produces the same assembly as before this patch modulo register numbers.

**Pull request merged**